python/cuml/common/array_sparse.py

```diff
@@ -0,0 +1,148 @@
+#
+# Copyright (c) 2020, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from cuml.common.import_utils import has_scipy
+
+import cupyx as cpx
+import numpy as np
+
+from cuml.common.memory_utils import with_cupy_rmm
+
+from cuml.common.input_utils import input_to_cuml_array
+
+if has_scipy():
+    import scipy.sparse
+
+
+class SparseCumlArray:
+    """
+    SparseCumlArray abstracts sparse array allocations. This will
+    accept either a Scipy or Cupy sparse array and construct CumlArrays
+    out of the underlying index and data arrays. Currently, this class
+    only supports the CSR array format and input in any other sparse
+    format will be converted to CSR.
+
+    Parameters
+    ----------
+
+    data : scipy.sparse.spmatrix or cupyx.scipy.sparse.spmatrix
+        A Scipy or Cupy sparse matrix
+    dtype : data-type, optional
+        Any object that can be interpreted as a numpy or cupy data type.
+
+    Attributes
+    ----------
+
+    indptr : CumlArray
+        Compressed row index array
+    indices : CumlArray
+        Column indices array
+    data : CumlArray
+        Data array
+    dtype : dtype
+        Data type of data array
+    shape : tuple of ints
+        Shape of the array
+    nnz : int
+        Number of nonzeros in underlying arrays
+    has_sorted_indices : bool
+        Whether column indices and data are sorted by column
+    """
+
+    @with_cupy_rmm
+    def __init__(self, data=None, dtype=None):
+        if not cpx.scipy.sparse.isspmatrix(data) and \
+                not (has_scipy() and scipy.sparse.isspmatrix(data)):
+            raise ValueError("A sparse matrix is expected as input. "
+                             "Received %s" % type(data))
+
+        data = data.tocsr()  # currently only CSR is supported
```

Comment:
are CSR matrices read-only in cupy? otherwise it seems like CumlArray allows mutation of the input data when given a CSR input but is an independent copy when given non-CSR, so behavior might be a little surprising then

Reply:
In Cupy, csr->csr conversion is a no-op (unless `copy=True`). Otherwise, a copy is going to have to be made because coo->csr and csc->csr can't reuse any of the arrays. 

This seems a little like the copy that CumlArray makes when it has to cast to a different dtype. We can issue a warning when the input is not CSR. Would that work?